draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1900 @@
+library unless certificates are already provided or the user has chosen to
+disable this behavior through a CLI option.
+
+SystemD Integration
+~~~~~~~~~~~~~~~~~~~
+
+Unless it is explicitly overridden by the configuration, whenever the Controller
+is run as a SystemD service, it will use SystemD to spawn all other Celery
+components.
+
+Celery will provide the required services for such a deployment.
+
+The Controller will use the `sd_notify`_ protocol to announce when the cluster
+is fully operational.
+
+The user must configure the list of hosts the controller will manage and ensure
+SSH communication between the Controller's host and the other hosts is possible.
+
+Other Integrations
+~~~~~~~~~~~~~~~~~~
+
+Celery may be run in Kubernetes, Swarm, Mesos, Nomad or any other container
+scheduler.
+
+Users may provide their own integrations with the Foreman which allows them to
+create and manage the different Celery components in a way that is native to the
+container scheduler.
+
+The Controller may also manage the lifecycle of the :term:`Message Broker` if
+the user wishes to do so.
+
+Such an integration may be provided by the user as well.
+
+Scheduler
++++++++++
+
+The scheduler is responsible for managing the scheduling of tasks for execution
+on a cluster of workers.
+
+The scheduler calculates the amount of tasks to be executed in any given time
+in order to make cluster wide decisions when autoscaling workers or increasing
+concurrency for an existing worker.
+
+The scheduler is aware when tasks should no longer be executed due to manual
+intervention or a circuit breaker trip. To do so, it commands the router to
+avoid consuming the task or rejecting it.
+
+Concurrency Limitations
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Not all :ref:`draft/celery-5-high-level-architecture:Tasks` are born equal.
+Some tasks require more resources than others, some may only be executed once
+at a time due to a business requirement,
+other tasks may be executed only once per user at a time to avoid data corruption.
+At times, some tasks should not be executed at all.
+
+The Scheduler is responsible for limiting the concurrency of such tasks.
+
+A task's concurrency may be limited per worker or globally across all workers
+depending on the requirements.
+In case there are tasks which are limited globally, an external data store is required.
+
+If a :ref:`task <draft/celery-5-high-level-architecture:Tasks>` is rate limited
+any concurrency limitations are ignored.
+
+There are multiple types of limits the user can impose on a task's concurrency:
+
+* **Fixed Limit**: A task can only be run at a maximum concurrency of a fixed number.
+  This strategy is used when there is a predetermined limit on the number of
+  concurrent tasks of the same type either because of lack of computing resources
+  or due to business requirements.
+* **Range**: A task can only be run at a maximum concurrency of a calculated limit
```

Comment:
`run -> ran`

Reply:
English is weirder than Perl.
https://www.quora.com/Which-is-more-grammatically-correct-was-run-or-was-ran

Reply:
I see, sorry for making the wrong suggestion :smile: !